src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -894,27 +907,139 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
   return next->delete_obj(dpp, y, flags);
 }
 
+
 int D4NFilterWriter::prepare(optional_yield y) 
 {
+  startTime = time(NULL);
   if (driver->get_cache_driver()->delete_data(save_dpp, obj->get_key().get_oid(), y) < 0) 
     ldpp_dout(save_dpp, 10) << "D4NFilterWriter::" << __func__ << "(): CacheDriver delete_data method failed." << dendl;
-
-  return next->prepare(y);
+  d4n_writecache = g_conf()->d4n_writecache_enabled;
+  if (d4n_writecache == false){
+    ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): calling next iterate" << dendl;
+    return next->prepare(y);
+  }
+  else
+    return 0;
 }
 
 int D4NFilterWriter::process(bufferlist&& data, uint64_t offset)
 {
-  /*
-  int append_dataReturn = driver->get_cache_driver()->append_data(save_dpp, obj->get_key().get_oid(), 
-  								    data, y);
+    bufferlist bl = data;
+    off_t bl_len = bl.length();
+    off_t ofs = offset;
+    bool dirty = true;
+    rgw::d4n::CacheBlock block, existing_block;
+    /*
+    std::stringstream s;
+    utime_t ut(obj->get_mtime());
+    ut.gmtime(s);
+    std::string creationTime = s.str();
+    */
+    auto creationTime = startTime;
 
-  if (append_dataReturn < 0) {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache append data operation failed." << dendl;
-  } else {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache append data operation succeeded." << dendl;
-  }*/
+    auto version = obj->get_instance();
+
+    std::string prefix;
+    if (version.empty()) { //for versioned objects, get_oid() returns an oid with versionId added
+      prefix =obj->get_bucket()->get_name() + "_" + obj->get_key().get_oid();
+    } else {
+      prefix = obj->get_bucket()->get_name() + "_" + version + "_" + obj->get_key().get_oid();
+    }
+
+    rgw::d4n::BlockDirectory* blockDir = driver->get_block_dir();
 
-  return next->process(std::move(data), offset);
+    block.hostsList.push_back(blockDir->cct->_conf->rgw_local_cache_address); 
+    block.cacheObj.bucketName = obj->get_bucket()->get_name();
+    block.cacheObj.objName = obj->get_key().get_oid();
+    block.cacheObj.dirty = dirty;
+    existing_block.cacheObj.objName = block.cacheObj.objName;
+    existing_block.cacheObj.bucketName = block.cacheObj.bucketName;
+
+
+    int ret = 0;
+
+      if (d4n_writecache == false){
+	
+        std::string oid = prefix + "_" + std::to_string(ofs)+ "_" + std::to_string(bl_len);
+        block.size = bl.length();
+        block.blockID = ofs;
+        block.dirty = false; //writing to the backend
+    	dirty = false;
+	
+  	ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): calling next process" << dendl;
+	ret = next->process(std::move(data), offset);
```

Comment:
is next->process writing to the storage backend? if `d4n_writecache` is false, why is the policy data structure being updated here? I understand that directory needs to be updated to for dirty/ non-dirty blocks. Policy data structure entries track blocks/ objects that are stored in the cache backend.

Reply:
Good point. It should be removed.